integration_tests/e2e/overview/goal/completeGoal.cy.ts

```diff
@@ -0,0 +1,152 @@
+
+  beforeEach(() => {
+    cy.task('reset')
+    cy.task('stubSignInAsUserWithEditAuthority')
+    cy.task('stubAuthUser')
+    cy.task('stubGetHeaderComponent')
+    cy.task('stubGetFooterComponent')
+    cy.task('stubPrisonerList')
+    cy.task('stubCiagInductionList')
+    cy.task('stubActionPlansList')
+    cy.task('getPrisonerById')
+    cy.task('stubGetInduction')
+    cy.task('getActionPlan')
+    cy.task('stubLearnerProfile')
+    cy.task('stubLearnerEducation')
+    cy.task('archiveGoal')
+    cy.task('stubGetAllPrisons')
+  })
+
+  it('should be able to navigate directly to complete goal page', () => {
+    // Given
+    cy.signIn()
+
+    // When
+    cy.visit(`/plan/${prisonNumber}/goals/${goalReference}/complete`)
+
+    // Then
+    const completeGoalPage = Page.verifyOnPage(CompleteGoalPage)
+    completeGoalPage.isForGoal(goalReference)
+  })
+
+  it('should be able to navigate to complete goals page from overview', () => {
+    // Given
+    cy.signIn()
+
+    cy.visit(`/plan/${prisonNumber}/view/overview`)
+    const overviewPage = Page.verifyOnPage(OverviewPage)
+
+    // When
+    const goalsPage = overviewPage.clickViewInProgressGoalsButton()
+    const completeOrArchivePage = goalsPage //
+      .checkOnInProgressGoalsTab()
+      .clickCompleteOrArchiveButtonForGoal(goalReference)
+    Page.verifyOnPage(CompleteOrArchiveGoalPage)
+    completeOrArchivePage.selectOption(CompleteOrArchiveGoalValue.COMPLETE).submitPage()
+
+    // Then
+    Page.verifyOnPage(CompleteGoalPage)
+  })
+
+  it('should ask for confirmation and return to overview if choosing no', () => {
+    // Given
+    cy.signIn()
+    cy.visit(`/plan/${prisonNumber}/goals/${goalReference}/complete`)
+    const completeGoalPage = Page.verifyOnPage(CompleteGoalPage)
+    completeGoalPage //
+      .clickNo()
+    // Then
+    Page.verifyOnPage(OverviewPage)
+    cy.wiremockVerifyNoInteractions(
+      putRequestedFor(urlEqualTo(`/action-plans/${prisonNumber}/goals/${goalReference}/complete`)),
+    )
+  })
+
+  it('Should be able to complete a goal successfully without entering the optional notes', () => {
+    // Given
+    cy.task('completeGoal', { prisonNumber, goalReference })
+    cy.signIn()
+    cy.visit(`/plan/${prisonNumber}/view/goals`)
+    const goalsPage = Page.verifyOnPage(GoalsPage)
+
+    // When
+    const completeOrArchiveGoalPage = goalsPage //
+      .hasArchivedGoalsDisplayed()
+      .hasNumberOfArchivedGoals(2)
+      .clickCompleteOrArchiveButtonForGoal(goalReference)
+
+    Page.verifyOnPage(CompleteOrArchiveGoalPage)
+
+    completeOrArchiveGoalPage //
+      .selectOption(CompleteOrArchiveGoalValue.COMPLETE)
+      .submitPage()
+
+    const completeGoalPage = Page.verifyOnPage(CompleteGoalPage)
+
+    completeGoalPage //
+      .clearNotes()
+      .clickYes()
+    // Then
+    Page.verifyOnPage(OverviewPage) //
+      .hasSuccessMessage('Goal Completed')
+      .hasNumberOfCompletedGoals(1)
```

Comment:
I would remove this part of the assertion (`hasNumberOfCompletedGoals`) because by making this assertion here in the `// Then` we are suggesting that some state has changed via the `// When` block (and we are checking that it has)
The reality is that because we use wiremock to mock our API calls, no state is ever changed. So even though the impl called the API to complete the goal, and we can assert that via the `cy.wiremockVerify` call, the actual state of data has not changed, so the rendering of the Overview page wont have changed from before the `// When` to after it.
If that makes sense?

Reply:
not sure - I looked at what the archive goal test did and did something similar. Although you are right in that the archive test checks before and after that the number of archived goals is 2 - so bit pointless really